src/appleseed/renderer/kernel/rendering/masterrenderer.cpp

```diff
@@ -75,416 +80,563 @@ namespace renderer
+        m_texture_system->attribute("automip", 0);
+        m_texture_system->attribute("accept_untiled", 1);
+        m_texture_system->attribute("accept_unmipped", 1);
+        m_texture_system->attribute("gray_to_rgb", 1);
+        m_texture_system->attribute("latlong_up", "y");
+        m_texture_system->attribute("flip_t", 1);
+
+        m_renderer_services = new RendererServices(
+            m_project,
+            reinterpret_cast<OIIO::TextureSystem&>(*m_texture_system));
+
+        RENDERER_LOG_DEBUG("creating osl shading system...");
+        m_shading_system = OSLShadingSystemFactory::create(
+            m_renderer_services,
+            m_texture_system,
+            m_error_handler);
+        m_shading_system->attribute("lockgeom", 1);
+        m_shading_system->attribute("colorspace", "Linear");
+        m_shading_system->attribute("commonspace", "world");
+        m_shading_system->attribute("statistics:level", 1);
+        m_shading_system->attribute(
+            "raytypes",
+            OSL::TypeDesc(
+                OSL::TypeDesc::STRING,
+                static_cast<int>(VisibilityFlags::Count)),
+            VisibilityFlags::Names);
+    #ifndef NDEBUG
+        m_shading_system->attribute("compile_report", 1);
+        m_shading_system->attribute("countlayerexecs", 1);
+        m_shading_system->attribute("clearmemory", 1);
+    #endif
+
+        // Register appleseed's closures into OSL's shading system.
+        register_closures(*m_shading_system);
+    }
 
-    delete impl->m_serial_tile_callback_factory;
-    delete impl->m_serial_renderer_controller;
-    delete impl;
-}
+    ~Impl()
+    {
+        if (m_display)
+            m_display->close();
 
-MasterRenderer::RenderingResult MasterRenderer::render()
-{
-    // Initialize thread-local variables.
-    Spectrum::set_mode(get_spectrum_mode(m_params));
+        delete m_serial_tile_callback_factory;
+        delete m_serial_renderer_controller;
 
-    RenderingResult result;
+        RENDERER_LOG_DEBUG("destroying osl shading system...");
+        m_project.get_scene()->release_optimized_osl_shader_groups();
+        m_shading_system->release();
+        delete m_renderer_services;
 
-    try
+        const string stats = m_texture_system->getstats();
+        const string modified_stats = prefix_all_lines(trim_both(stats), "oiio: ");
+        RENDERER_LOG_DEBUG("%s", modified_stats.c_str());
+
+        RENDERER_LOG_DEBUG("destroying oiio texture system...");
+        m_texture_system->release();
+        delete m_error_handler;
+    }
+
+    // Initialize OSL's shading system.
+    bool initialize_osl_shading_system(
+        TextureStore&               texture_store,
+        foundation::IAbortSwitch&   abort_switch)
     {
-        Stopwatch<DefaultWallclockTimer> stopwatch;
-        stopwatch.start();
+        // Initialize OIIO.
+        const ParamArray& params = m_params.child("texture_store");
```

Comment:
This variable is confusing. AFAICT, it's used only once so I would suggest inlining it.